File: sdks/react-native/src/UdeneClient.ts

import axios, { AxiosInstance } from 'axios';

/**
 * Configuration options for the UdeneClient
 */
export interface UdeneClientConfig {
  /**
   * Your Udene API key (required)
   */
  apiKey: string;
  
  /**
   * Custom API base URL (optional)
   * @default "https://api.udene.com/v1"
   */
  baseURL?: string;
  
  /**
   * Device platform (optional, auto-detected if not provided)
   * @default auto-detected from React Native
   */
  platform?: string;
}

/**
 * Interaction data for tracking user behaviors
 */
export interface InteractionData {
  /**
   * User identifier
   */
  userId?: string;
  
  /**
   * Type of action performed
   */
  action: string;
  
  /**
   * Additional data about the interaction
   */
  metadata?: Record<string, any>;
}

/**
 * Main client for Udene fraud detection services
 */
export class UdeneClient {
  private client: AxiosInstance;
  private platform: string;

  /**
   * Creates a new UdeneClient instance
   * 
   * @param config - Configuration options
   */
  constructor(config: UdeneClientConfig) {
    if (!config.apiKey) {
      throw new Error('API key is required');
    }

    this.platform = config.platform || 'react-native';
    
    this.client = axios.create({
      baseURL: config.baseURL || 'https://api.udene.com/v1',
      headers: {
        'Authorization': `Bearer ${config.apiKey}`,
        'Content-Type': 'application/json',
        'X-Udene-SDK': `react-native-sdk/1.0.0`,
        'X-Udene-Platform': this.platform
      }
    });
  }

  /**
   * Get fraud detection metrics for the current session/user
   * 
   * @returns Promise with metrics data
   */
  async getMetrics(): Promise<any> {
    try {
      const response = await this.client.get('/metrics');
      return response.data;
    } catch (error) {
      this.handleError(error);
      throw error;
    }
  }

  /**
   * Analyze Business Email Compromise threats
   * 
   * @param emailData - Email data to analyze
   * @returns Promise with analysis results
   */
  async analyzeBEC(emailData: any): Promise<any> {
    try {
      const response = await this.client.post('/analyze/bec', emailData);
      return response.data;
    } catch (error) {
      this.handleError(error);
      throw error;
    }
  }

  /**
   * Track user interactions for fraud analysis
   * 
   * @param data - Interaction data
   * @returns Promise with tracking confirmation
   */
  async trackInteraction(data: InteractionData): Promise<any> {
    try {
      const response = await this.client.post('/track', data);
      return response.data;
    } catch (error) {
      this.handleError(error);
      throw error;
    }
  }

  /**
   * Get user activity data
   * 
   * @returns Promise with user activity data
   */
  async getActivity(): Promise<any> {
    try {
      const response = await this.client.get('/activity');
      return response.data;
    } catch (error) {
      this.handleError(error);
      throw error;
    }
  }

  /**
   * Handle API errors
   * 
   * @param error - Error object from API request
   * @private
   */
  private handleError(error: any): void {
    if (axios.isAxiosError(error)) {
      const serverError = error.response?.data?.error;
      if (serverError) {
        console.error(`Udene API Error: ${serverError.message || 'Unknown error'}`);
      }
    }
  }
}

File: sdks/ios/Sources/UdeneSDK/UdeneClient.swift

import Foundation
#if canImport(FoundationNetworking)
import FoundationNetworking
#endif
#if canImport(UIKit)
import UIKit
#endif

/// SDK Version
public let kUdeneSDKVersion = "1.0.3"

/// Client for interacting with the Udene Fraud Detection API
public class UdeneClient {
    /// The SDK version
    public static var sdkVersion: String {
        return kUdeneSDKVersion
    }

    /// Determines if the code is running on an Apple platform
    public static func isApplePlatform() -> Bool {
        #if os(macOS) || os(iOS) || os(tvOS) || os(watchOS)
        return true
        #else
        return false
        #endif
    }

    /// Determines if the code is running on Windows
    public static func isWindowsPlatform() -> Bool {
        #if os(Windows)
        return true
        #else
        return false
        #endif
    }

    /// Determines if the code is running on Linux
    public static func isLinuxPlatform() -> Bool {
        #if os(Linux)
        return true
        #else
        return false
        #endif
    }

    /// Returns the current platform name
    public static func platformName() -> String {
        #if os(iOS)
        return "iOS"
        #elseif os(macOS)
        return "macOS"
        #elseif os(tvOS)
        return "tvOS"
        #elseif os(watchOS)
        return "watchOS"
        #elseif os(Windows)
        return "Windows"
        #elseif os(Linux)
        return "Linux"
        #else
        return "Unknown"
        #endif
    }
    /// API Key for authentication
    private let apiKey: String
    
    /// Base URL for API requests
    private let baseURL: URL
    
    /// URLSession for network requests
    private let session: URLSession
    
    /// User-Agent string for requests
    private let userAgent: String
    
    /// Initializes a new UdeneClient
    /// - Parameters:
    ///   - apiKey: Your Udene API key
    ///   - baseURL: Optional custom base URL for the API
    ///   - session: Optional custom URLSession
    public init(
        apiKey: String,
        baseURLString: String = "https://udene.net/v1",
        session: URLSession = .shared
    ) {
        self.apiKey = apiKey
        
        guard let url = URL(string: baseURLString) else {
            fatalError("Invalid base URL: \(baseURLString)")
        }
        
        self.baseURL = url
        self.session = session
        
        var appVersion = kUdeneSDKVersion
        #if os(iOS)
            if let bundleVersion = Bundle.main.infoDictionary?["CFBundleShortVersionString"] as? String {
                appVersion = bundleVersion
            }
            self.userAgent = "UdeneSDK/\(kUdeneSDKVersion) iOS/\(UIDevice.current.systemVersion) App/\(appVersion)"
        #elseif os(macOS)
            self.userAgent = "UdeneSDK/\(kUdeneSDKVersion) macOS Swift/5.7"
        #elseif os(Windows)
            self.userAgent = "UdeneSDK/\(kUdeneSDKVersion) Windows Swift/5.7"
        #elseif os(Linux)
            self.userAgent = "UdeneSDK/\(kUdeneSDKVersion) Linux Swift/5.7"
        #else
            self.userAgent = "UdeneSDK/\(kUdeneSDKVersion) Swift/5.7"
        #endif
    }
    
    /// Gets fraud metrics
    /// - Parameter completion: Completion handler with Result
    public func getMetrics(completion: @escaping (Result<Metrics, UdeneError>) -> Void) {
        let url = baseURL.appendingPathComponent("metrics")
        var request = URLRequest(url: url)
        
        addHeaders(to: &request)
        
        let task = session.dataTask(with: request) { data, response, error in
            if let error = error {
                completion(.failure(.networkError(error)))
                return
            }
            
            guard let httpResponse = response as? HTTPURLResponse else {
                completion(.failure(.invalidResponse))
                return
            }
            
            self.handleResponse(data: data, response: httpResponse, completion: completion)
        }
        
        task.resume()
    }
    
    /// Tracks a user interaction
    /// - Parameters:
    ///   - userId: User identifier
    ///   - action: Action name
    ///   - metadata: Additional metadata
    ///   - completion: Completion handler with Result
    public func trackInteraction(
        userId: String,
        action: String,
        metadata: [String: Any]? = nil,
        completion: @escaping (Result<InteractionResponse, UdeneError>) -> Void
    ) {
        let url = baseURL.appendingPathComponent("track")
        var request = URLRequest(url: url)
        request.httpMethod = "POST"
        
        addHeaders(to: &request)
        
        var body: [String: Any] = [
            "userId": userId,
            "action": action
        ]
        
        if let metadata = metadata {
            body["metadata"] = metadata
        }
        
        do {
            request.httpBody = try JSONSerialization.data(withJSONObject: body)
        } catch {
            completion(.failure(.encodingError(error)))
            return
        }
        
        let task = session.dataTask(with: request) { data, response, error in
            if let error = error {
                completion(.failure(.networkError(error)))
                return
            }
            
            guard let httpResponse = response as? HTTPURLResponse else {
                completion(.failure(.invalidResponse))
                return
            }
            
            self.handleResponse(data: data, response: httpResponse, completion: completion)
        }
        
        task.resume()
    }
    
    /// Gets activity data
    /// - Parameter completion: Completion handler with Result
    public func getActivity(completion: @escaping (Result<Activity, UdeneError>) -> Void) {
        let url = baseURL.appendingPathComponent("activity")
        var request = URLRequest(url: url)
        
        addHeaders(to: &request)
        
        let task = session.dataTask(with: request) { data, response, error in
            if let error = error {
                completion(.failure(.networkError(error)))
                return
            }
            
            guard let httpResponse = response as? HTTPURLResponse else {
                completion(.failure(.invalidResponse))
                return
            }
            
            self.handleResponse(data: data, response: httpResponse, completion: completion)
        }
        
        task.resume()
    }
    
    // MARK: - Private helpers
    
    private func addHeaders(to request: inout URLRequest) {
        request.addValue("Bearer \(apiKey)", forHTTPHeaderField: "Authorization")
        request.addValue("application/json", forHTTPHeaderField: "Content-Type")
        request.addValue(userAgent, forHTTPHeaderField: "User-Agent")
        request.addValue(kUdeneSDKVersion, forHTTPHeaderField: "X-Client-Version")
        request.addValue(UdeneClient.platformName().lowercased(), forHTTPHeaderField: "X-SDK-Type")
    }
    
    private func handleResponse<T: Decodable>(
        data: Data?,
        response: HTTPURLResponse,
        completion: @escaping (Result<T, UdeneError>) -> Void
    ) {
        switch response.statusCode {
        case 200...299:
            guard let data = data else {
                completion(.failure(.noData))
                return
            }
            
            do {
                let decoder = JSONDecoder()
                decoder.keyDecodingStrategy = .convertFromSnakeCase
                let result = try decoder.decode(T.self, from: data)
                completion(.success(result))
            } catch {
                completion(.failure(.decodingError(error)))
            }
            
        case 401:
            completion(.failure(.unauthorized))
            
        case 429:
            let retryAfter = response.value(forHTTPHeaderField: "Retry-After")
                .flatMap { Int($0) } ?? 60
            completion(.failure(.rateLimit(retryAfter: retryAfter)))
            
        case 400...499:
            completion(.failure(.clientError(statusCode: response.statusCode)))
            
        case 500...599:
            completion(.failure(.serverError(statusCode: response.statusCode)))
            
        default:
            completion(.failure(.unknown))
        }
    }
}
